src/ApiClient/util/AuthController.ts

```diff
@@ -0,0 +1,232 @@
+            return;
+        }
+        return this.token;
+    }
+
+    /**
+     * Get token without handling validation.
+     */
+    public getTokenUnsafe() {
+        return this.token;
+    }
+
+    /**
+     * Authenticate using the cached credential information
+     */
+    public async authenticateCached() {
+        if (!this._credentials) {
+            return new InternalStatus<undefined, LoginErrors>({
+                code: StatusCode.ERROR,
+                error: new InternalError(ErrorCode.LOGIN_NOCREDS, { void: true })
+            });
+        }
+        return await this.authenticate(this._credentials);
+    }
+
+    /**
+     * Attempt to authenticate using the credentials.
+     * This **will not** return token from status.
+     * @param credential The credentials, either password/username or oauth.
+     * @param cacheCredential Should we cache the credentials for later login?
+     */
+    public async authenticate(credential: ICredentials, cacheCredential = true) {
+        if (!credential) {
+            return new InternalStatus<undefined, LoginErrors>({
+                code: StatusCode.ERROR,
+                error: new InternalError(ErrorCode.LOGIN_NOCREDS, { void: true })
+            });
+        }
+        // debounce (iffy about this)
+        if (this.loggingIn) {
+            return new InternalStatus<undefined, LoginErrors>({
+                code: StatusCode.ERROR,
+                error: new InternalError(ErrorCode.LOGIN_LOGGING_IN, { void: true })
+            });
+        }
+        if (cacheCredential) {
+            this._credentials = credential;
+        }
+
+        let response;
+        try {
+            if (credential.type == CredentialsType.Password) {
+                response = await this.apiClient.homeControllerCreateToken({
+                    auth: {
+                        username: credential.userName,
+                        password: credential.password
+                    }
+                });
+            } else if (credential.type == CredentialsType.OAuth) {
+                response = await this.apiClient.homeControllerCreateToken({
+                    headers: {
+                        OAuthProvider: credential.provider,
+                        Authorization: `OAuth ${credential.token}`
+                    }
+                });
+            }
+        } catch (err) {
+            return new InternalStatus<undefined, LoginErrors>({
+                code: StatusCode.ERROR,
+                error: new InternalError<LoginErrors>(ErrorCode.APP_FAIL, { jsError: err as Error })
+            });
+        } finally {
+            this.loggingIn = false;
+        }
+
+        if (!response || !response.data) {
+            return new InternalStatus<undefined, LoginErrors>({
+                code: StatusCode.ERROR,
+                error: new InternalError(ErrorCode.HTTP_NOT_ACCEPTABLE, { void: true })
+            });
+        }
+
+        if (response.status !== 200) {
```

Comment:
Uh, why indent shit that doesn't need to be indented, what's wrong with case 200: ?